File: 19. Remove Nth Node From End of List.cpp
//
//  19. Remove Nth Node From End of List.h
//  Project 1
//
//  Created by Jonny Kong on 2/22/16.
//  Copyright © 2016 Jonny Kong. All rights reserved.
//

ListNode* removeNthFromEnd(ListNode* head, int n) {
    
    int i = 0;
    
    ListNode* p = head;
    
    while(p -> next){
        
        p = p -> next;
        
        ++i;
        
    }
    
    if(i == 0) return NULL; //only one node
    
    else if(i == 1){
        
        if(n == 1){
            
            head -> next = NULL;
            
            return head;
            
        }
        
        else return head -> next;
        
    }
    
    if(i == n - 1) return head -> next;
    
    p = head;
    
    for(int j=0; j < i - n; ++j){
        
        p = p -> next;
        
    }
    
    p -> next = p -> next -> next;
    
    return head;
    
}

File: 19_Remove_Nth_Node_From_End_of_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode p = head;
        int listLength = 1;

        /* Get length of list */
        while(p.next != null) {
            p = p.next;
            listLength++;
        }

        /* If the first node is to be removed */
        if(listLength == n)
            return head.next;

        /* Move to the node preceding the one to be removed */
        p = head;
        for(int i = 0; i < listLength - n - 1; ++i)
            p = p.next;

        /* Remove this node */
        p.next = p.next.next;
        return head;
    }
}



